validator/server_common/machine_locator.go

```diff
@@ -39,31 +38,65 @@ func NewMachineLocator(rootPath string) (*MachineLocator, error) {
 		}
 		workPath1 := filepath.Join(workDir, "machines")
 		workPath2 := filepath.Join(filepath.Join(workDir, "target"), "machines")
-		places = append(places, workPath1)
-		places = append(places, workPath2)
+		dirs = append(dirs, workPath1)
+		dirs = append(dirs, workPath2)
 
 		// Check above the executable: <binary> => ../../machines
 		execfile, err := os.Executable()
 		if err != nil {
 			return nil, err
 		}
 		execPath := filepath.Join(filepath.Dir(filepath.Dir(execfile)), "machines")
-		places = append(places, execPath)
+		dirs = append(dirs, execPath)
 	}
 
-	for _, place := range places {
-		if _, err := os.Stat(place); err == nil {
-			var latestModuleRoot common.Hash
-			latestModuleRootPath := filepath.Join(place, "latest", "module-root.txt")
-			fileBytes, err := os.ReadFile(latestModuleRootPath)
-			if err == nil {
-				s := strings.TrimSpace(string(fileBytes))
-				latestModuleRoot = common.HexToHash(s)
+	var (
+		moduleRoots      = make(map[common.Hash]bool)
+		latestModuleRoot common.Hash
+	)
+
+	for _, dir := range dirs {
+		fInfo, err := os.Stat(dir)
+		if err != nil {
+			log.Warn("Getting file info", "error", err)
+			continue
+		}
+		if !fInfo.IsDir() {
+			// Skip files that are not directories.
+			continue
+		}
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			log.Warn("Reading directory", "dir", dir, "error", err)
+		}
+		for _, file := range files {
+			mrFile := filepath.Join(dir, file.Name(), "module-root.txt")
+			if _, err := os.Stat(mrFile); errors.Is(err, os.ErrNotExist) {
+				// Skip if module-roots file does not exist.
+				continue
+			}
+			mrContent, err := os.ReadFile(mrFile)
+			if err != nil {
+				log.Warn("Reading module roots file", "file path", mrFile, "error", err)
+				continue
+			}
+			moduleRoot := common.HexToHash(strings.TrimSpace(string(mrContent)))
+			moduleRoots[moduleRoot] = true
+			if file.Name() == "latest" {
+				latestModuleRoot = moduleRoot
+				rootPath = dir
 			}
```

Comment:
Only include dirs if their name is either "latest" or the wasmMduleRoot - otherwise locator will not be able to find it 
see GetMachinePath.

Reply:
Isn't that exactly what I'm doing already ? Iterating all folders, checking if there's `module-root.txt` then reading file content (which is same as directory name if it's not "latest"), then adding to moduleRoots.

Or am I missing something ?

Reply:
if you have a file "foobar/module-root.txt" - you cannot use this module-root because the locator won't be able to find "foobar" dir later. We can change that in GetMachinePath.. but easier to just not include it in the list.